File: examples/paper/src/main/java/me/devnatan/inventoryframework/runtime/view/SimplePagination.java
package me.devnatan.inventoryframework.runtime.view;

import me.devnatan.inventoryframework.View;
import me.devnatan.inventoryframework.ViewConfigBuilder;
import me.devnatan.inventoryframework.component.Pagination;
import me.devnatan.inventoryframework.context.Context;
import me.devnatan.inventoryframework.context.RenderContext;
import me.devnatan.inventoryframework.runtime.ExampleUtil;
import me.devnatan.inventoryframework.state.State;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

public class SimplePagination extends View {

    private final State<Pagination> state =
            lazyPaginationState(() -> ExampleUtil.getRandomItems(123), (context, builder, index, value) -> {
                builder.withItem(value);
                builder.onClick((ctx) -> {
                    ctx.getPlayer().sendMessage("You clicked on item " + index);
                });
            });

    @Override
    public void onInit(@NotNull ViewConfigBuilder config) {
        config.cancelOnClick();
        config.size(3);
        config.title("Simple Pagination");
        config.layout("OOOOOOOOO", "OOOOOOOOO", "  P   N  ");
    }

    @Override
    public void onFirstRender(@NotNull RenderContext render) {
        ItemStack previousItem = ExampleUtil.displayItem(Material.ARROW, "Previous");
        ItemStack nextItem = ExampleUtil.displayItem(Material.ARROW, "Next");
        render.layoutSlot('P', previousItem)
                .displayIf(() -> state.get(render).canBack())
                .updateOnStateChange(state)
                .onClick((ctx) -> {
                    Pagination pagination = state.get(ctx);
                    pagination.back();
                    ctx.updateTitleForPlayer("Simple Pagination (" + pagination.currentPage() + ")");
                });
        render.layoutSlot('N', nextItem)
                .displayIf(() -> state.get(render).canAdvance())
                .updateOnStateChange(state)
                .onClick((ctx) -> {
                    Pagination pagination = state.get(ctx);
                    pagination.advance();
                    ctx.updateTitleForPlayer("Simple Pagination (" + pagination.currentPage() + ")");
                });

        render.lastSlot(new ItemStack(Material.ARROW))
                .displayIf(Context::canBack)
                .onClick(Context::back);
    }
}

File: examples/minestom/src/main/kotlin/me/devnatan/inventoryframework/runtime/view/SimplePagination.kt
package me.devnatan.inventoryframework.runtime.view

import me.devnatan.inventoryframework.View
import me.devnatan.inventoryframework.ViewConfigBuilder
import me.devnatan.inventoryframework.component.MinestomIemComponentBuilder
import me.devnatan.inventoryframework.component.Pagination
import me.devnatan.inventoryframework.context.Context
import me.devnatan.inventoryframework.context.RenderContext
import me.devnatan.inventoryframework.context.SlotClickContext
import me.devnatan.inventoryframework.runtime.ExampleUtil.displayItem
import me.devnatan.inventoryframework.runtime.ExampleUtil.getRandomItems
import me.devnatan.inventoryframework.state.State
import net.minestom.server.item.ItemStack
import net.minestom.server.item.Material

class SimplePagination : View() {
    private val state: State<Pagination> =
        lazyPaginationState(
            { _ -> getRandomItems(123).toMutableList() },
            { _: Context, builder: MinestomIemComponentBuilder, index: Int, value: ItemStack ->
                builder.withItem(value)
                builder.onClick { ctx: SlotClickContext ->
                    ctx.player.sendMessage(
                        "You clicked on item $index",
                    )
                }
            },
        )

    override fun onInit(config: ViewConfigBuilder) {
        config.cancelOnClick()
        config.size(3)
        config.title("Simple Pagination")
        config.layout("OOOOOOOOO", "OOOOOOOOO", "  P   N  ")
    }

    override fun onFirstRender(render: RenderContext) {
        val previousItem = displayItem(Material.ARROW, "Previous")
        val nextItem = displayItem(Material.ARROW, "Next")
        render.layoutSlot('P', previousItem)
            .displayIf({ ctx -> state[ctx].canBack() })
            .updateOnStateChange(state)
            .onClick { ctx: SlotClickContext -> state[ctx].back() }
        render.layoutSlot('N', nextItem)
            .displayIf({ ctx -> state[ctx].canAdvance() })
            .updateOnStateChange(state)
            .onClick { ctx: SlotClickContext -> state[ctx].advance() }
    }
}
